### ISSUE TITLE
Performance tips + code reduction suggestion

### ISSUE THREAD

**Issue by nowindowsowrking** [jquery/esprima#1640]
@mikesherov @ariya  After spending the day today to read the entire source code of Esprima, I found a few things worth mention. 

- `this.scanner` can be removed and replaced with a simple `this` if you let the `Parser` and `Tokenizer` extend the `Scanner class`. There will be no need to create a new instance.

- Tokens.  A dictonary lookup is slow, and if you turn the values into bit values instead, and do a bit value comparison. The performance will increase. (*tested it myself*)

- This code https://github.com/jquery/esprima/blob/master/src/token.ts#L14 should be refatored into an function returning an  switch statement where the bit numbers are the case values. This will make V8 happy, and improve performance.

And I noticed that the fromCodePoint function doesn't handle  `UTF-16 Decoding` correctly? I am probably mistaken (*it's the middle of the night now*), but I think a faster and improved version of that function could be something like this:

```js
fromCodePoint(cp: number): string {
    if (cp <= 0xFFFF) {
        return String.fromCharCode(cp);
    }
    cp -= 0x10000;
    return String.fromCharCode((cp >> 10) + 0xD800, (cp % 0x400) + 0xDC00);
}
```

**Comment by ariya** [comment]
Thanks @nowindowsowrking!

In order to make it easier to follow the progress and to ease future maintenance, for situation like this, I recommend the following (instead of putting everything in one big giant bucket):

To discuss general performance analysis and optimization idea, post it to the mailing-list: https://groups.google.com/forum/#!forum/esprima.

To get feedback on an particular performance improvement, open an issue with a specific title (not just the generic "faster this, faster that"). If there is a work-in-progress patch that leads to a non-negligible speed-up (`npm run benchmark`), create a pull request referring the said issue.

Looking forward to seeing your patches!

**Comment by nowindowsowrking** [comment]
@ariya Right now I do not think it will be right if I come with some detailed ideas about performance and improvements. This is because I am not so deep into this as you are. And there are many different things to keep track of which must fit together to get a streamlined performance boost without bottlenecks.

But because of the fact that I like Esprima and code quality, and also the fact that I am "hooked", I jumped into this now.

The first thing I do now is to develop my own prototype parser, which I will use as a basis for code analysis of Esprima before I come back with suggestions. I am sick these days - flu - so I have plenty of time.

As soon as I have a parser that is around 50% faster then Esprima, and works cross-browser as well on NodeJS I will come back to this. Give it 2-3 days.

**Comment by nowindowsowrking** [comment]
@ariya This seems to be possible results that you can archive with optimizing:

```js
var a, b, c;
Esprima x 383,202 ops/sec ±0.80% (85 runs sampled)
XXX x 33,760,619 ops/sec ±1.10% (88 runs sampled)
Fastest is XXX
```

**Comment by nowindowsowrking** [comment]
@ariya  @mikesherov I got an understanding now on how things works, and I can contribute with tips and ideas for improvements. Where is the channel for this?

After my findings I concluded that Esprimas performance can be increased with around 6 -8% without heavy modifications of current source code.

However I wrote my own parser the last 24 hours as an prototype to figure out the exact limit of performance we can expect from an ECMAScript parser.

Listed below is the end results, and this is also possible with Esprima, but require that we would need a chat somewhere.

```js
// tested
 var a = 1, b = 2;"
// result
Esprima 230,600 ops/sec ±2.63% (83 runs sampled)
XXX x 1,221,403 ops/sec ±2.31% (85 runs sampled)
Fastest is XXX

// tested
 "a || b && c | d ^ e & f == g < h >>> i + j * k"

// result
Esprima x 89,335 ops/sec ±0.79% (88 runs sampled)
XXX x 252,456 ops/sec ±1.21% (88 runs sampled)
Fastest is XXX
```
